File: book/petrol-pump/java/pump/Outputs.java
package pump;

import nz.sodium.*;

public class Outputs {
    private Outputs(
            Cell<Delivery> delivery,
            Cell<String> presetLCD,
            Cell<String> saleCostLCD,
            Cell<String> saleQuantityLCD,
            Cell<String> priceLCD1,
            Cell<String> priceLCD2,
            Cell<String> priceLCD3,
            Stream<Unit> sBeep,
            Stream<Sale> sSaleComplete) {
        this.delivery = delivery;
        this.presetLCD = presetLCD;
        this.saleCostLCD = saleCostLCD;
        this.saleQuantityLCD = saleQuantityLCD;
        this.priceLCD1 = priceLCD1;
        this.priceLCD2 = priceLCD2;
        this.priceLCD3 = priceLCD3;
        this.sBeep = sBeep;
        this.sSaleComplete = sSaleComplete;
    }

    public Outputs() {
        this.delivery = new Cell<Delivery>(Delivery.OFF);
        this.presetLCD = new Cell<String>("");
        this.saleCostLCD = new Cell<String>("");
        this.saleQuantityLCD = new Cell<String>("");
        this.priceLCD1 = new Cell<String>("");
        this.priceLCD2 = new Cell<String>("");
        this.priceLCD3 = new Cell<String>("");
        this.sBeep = new Stream<Unit>();
        this.sSaleComplete = new Stream<Sale>();
    }

    public final Cell<Delivery> delivery;
    public final Cell<String> presetLCD;
    public final Cell<String> saleCostLCD;
    public final Cell<String> saleQuantityLCD;
    public final Cell<String> priceLCD1;
    public final Cell<String> priceLCD2;
    public final Cell<String> priceLCD3;
    public final Stream<Unit> sBeep;
    public final Stream<Sale> sSaleComplete;

    public Outputs setDelivery(Cell<Delivery> delivery) {
        return new Outputs(delivery, presetLCD, saleCostLCD,
                saleQuantityLCD, priceLCD1, priceLCD2, priceLCD3, sBeep,
                sSaleComplete);
    }
    public Outputs setPresetLCD(Cell<String> presetLCD) {
        return new Outputs(delivery, presetLCD, saleCostLCD,
                saleQuantityLCD, priceLCD1, priceLCD2, priceLCD3, sBeep,
                sSaleComplete);
    }
    public Outputs setSaleCostLCD(Cell<String> saleCostLCD) {
        return new Outputs(delivery, presetLCD, saleCostLCD,
                saleQuantityLCD, priceLCD1, priceLCD2, priceLCD3, sBeep,
                sSaleComplete);
    }
    public Outputs setSaleQuantityLCD(Cell<String> saleQuantityLCD) {
        return new Outputs(delivery, presetLCD, saleCostLCD,
                saleQuantityLCD, priceLCD1, priceLCD2, priceLCD3, sBeep,
                sSaleComplete);
    }
    public Outputs setPriceLCD1(Cell<String> priceLCD1) {
        return new Outputs(delivery, presetLCD, saleCostLCD,
                saleQuantityLCD, priceLCD1, priceLCD2, priceLCD3, sBeep,
                sSaleComplete);
    }
    public Outputs setPriceLCD2(Cell<String> priceLCD2) {
        return new Outputs(delivery, presetLCD, saleCostLCD,
                saleQuantityLCD, priceLCD1, priceLCD2, priceLCD3, sBeep,
                sSaleComplete);
    }
    public Outputs setPriceLCD3(Cell<String> priceLCD3) {
        return new Outputs(delivery, presetLCD, saleCostLCD,
                saleQuantityLCD, priceLCD1, priceLCD2, priceLCD3, sBeep,
                sSaleComplete);
    }
    public Outputs setBeep(Stream<Unit> sBeep) {
        return new Outputs(delivery, presetLCD, saleCostLCD,
                saleQuantityLCD, priceLCD1, priceLCD2, priceLCD3, sBeep,
                sSaleComplete);
    }
    public Outputs setSaleComplete(Stream<Sale> sSaleComplete) {
        return new Outputs(delivery, presetLCD, saleCostLCD,
                saleQuantityLCD, priceLCD1, priceLCD2, priceLCD3, sBeep,
                sSaleComplete);
    }
}

File: book/petrol-pump/c#/PetrolPump/Outputs.cs
using Sodium.Frp;
using Sodium.Functional;

namespace PetrolPump
{
    public class Outputs
    {
        private Outputs(
            Cell<Delivery> delivery,
            Cell<string> presetLcd,
            Cell<string> saleCostLcd,
            Cell<string> saleQuantityLcd,
            Cell<string> priceLcd1,
            Cell<string> priceLcd2,
            Cell<string> priceLcd3,
            Stream<Unit> sBeep,
            Stream<Sale> sSaleComplete)
        {
            this.Delivery = delivery;
            this.PresetLcd = presetLcd;
            this.SaleCostLcd = saleCostLcd;
            this.SaleQuantityLcd = saleQuantityLcd;
            this.PriceLcd1 = priceLcd1;
            this.PriceLcd2 = priceLcd2;
            this.PriceLcd3 = priceLcd3;
            this.SBeep = sBeep;
            this.SSaleComplete = sSaleComplete;
        }

        public Outputs()
        {
            this.Delivery = Cell.Constant(PetrolPump.Delivery.Off);
            this.PresetLcd = Cell.Constant(string.Empty);
            this.SaleCostLcd = Cell.Constant(string.Empty);
            this.SaleQuantityLcd = Cell.Constant(string.Empty);
            this.PriceLcd1 = Cell.Constant(string.Empty);
            this.PriceLcd2 = Cell.Constant(string.Empty);
            this.PriceLcd3 = Cell.Constant(string.Empty);
            this.SBeep = Stream.Never<Unit>();
            this.SSaleComplete = Stream.Never<Sale>();
        }

        public Cell<Delivery> Delivery { get; }
        public Cell<string> PresetLcd { get; }
        public Cell<string> SaleCostLcd { get; }
        public Cell<string> SaleQuantityLcd { get; }
        public Cell<string> PriceLcd1 { get; }
        public Cell<string> PriceLcd2 { get; }
        public Cell<string> PriceLcd3 { get; }
        public Stream<Unit> SBeep { get; }
        public Stream<Sale> SSaleComplete { get; }

        public Outputs SetDelivery(Cell<Delivery> delivery)
        {
            return new Outputs(delivery, this.PresetLcd, this.SaleCostLcd,
                this.SaleQuantityLcd, this.PriceLcd1, this.PriceLcd2, this.PriceLcd3, this.SBeep,
                this.SSaleComplete);
        }

        public Outputs SetPresetLcd(Cell<string> presetLcd)
        {
            return new Outputs(this.Delivery, presetLcd, this.SaleCostLcd,
                this.SaleQuantityLcd, this.PriceLcd1, this.PriceLcd2, this.PriceLcd3, this.SBeep,
                this.SSaleComplete);
        }

        public Outputs SetSaleCostLcd(Cell<string> saleCostLcd)
        {
            return new Outputs(this.Delivery, this.PresetLcd, saleCostLcd,
                this.SaleQuantityLcd, this.PriceLcd1, this.PriceLcd2, this.PriceLcd3, this.SBeep,
                this.SSaleComplete);
        }

        public Outputs SetSaleQuantityLcd(Cell<string> saleQuantityLcd)
        {
            return new Outputs(this.Delivery, this.PresetLcd, this.SaleCostLcd,
                saleQuantityLcd, this.PriceLcd1, this.PriceLcd2, this.PriceLcd3, this.SBeep,
                this.SSaleComplete);
        }

        public Outputs SetPriceLcd1(Cell<string> priceLcd1)
        {
            return new Outputs(this.Delivery, this.PresetLcd, this.SaleCostLcd,
                this.SaleQuantityLcd, priceLcd1, this.PriceLcd2, this.PriceLcd3, this.SBeep,
                this.SSaleComplete);
        }

        public Outputs SetPriceLcd2(Cell<string> priceLcd2)
        {
            return new Outputs(this.Delivery, this.PresetLcd, this.SaleCostLcd,
                this.SaleQuantityLcd, this.PriceLcd1, priceLcd2, this.PriceLcd3, this.SBeep,
                this.SSaleComplete);
        }

        public Outputs SetPriceLcd3(Cell<string> priceLcd3)
        {
            return new Outputs(this.Delivery, this.PresetLcd, this.SaleCostLcd,
                this.SaleQuantityLcd, this.PriceLcd1, this.PriceLcd2, priceLcd3, this.SBeep,
                this.SSaleComplete);
        }

        public Outputs SetBeep(Stream<Unit> sBeep)
        {
            return new Outputs(this.Delivery, this.PresetLcd, this.SaleCostLcd,
                this.SaleQuantityLcd, this.PriceLcd1, this.PriceLcd2, this.PriceLcd3, sBeep,
                this.SSaleComplete);
        }

        public Outputs SetSaleComplete(Stream<Sale> sSaleComplete)
        {
            return new Outputs(this.Delivery, this.PresetLcd, this.SaleCostLcd,
                this.SaleQuantityLcd, this.PriceLcd1, this.PriceLcd2, this.PriceLcd3, this.SBeep,
                sSaleComplete);
        }
    }
}